src/java.base/share/classes/java/lang/invoke/StringConcatFactory.java

```diff
@@ -1346,4 +1368,149 @@ public static MethodHandle makeConcatWithTemplateGetters(
+        static final ClassDesc STRING_BUILDER = ClassDesc.of("java.lang.StringBuilder");
+        static final ClassFileDumper DUMPER;
+        static final MethodTypeDesc APPEND_BOOLEAN_TYPE = MethodTypeDesc.ofDescriptor("(Z)Ljava/lang/StringBuilder;");
+        static final MethodTypeDesc APPEND_CHAR_TYPE = MethodTypeDesc.ofDescriptor("(C)Ljava/lang/StringBuilder;");
+        static final MethodTypeDesc APPEND_DOUBLE_TYPE = MethodTypeDesc.ofDescriptor("(D)Ljava/lang/StringBuilder;");
+        static final MethodTypeDesc APPEND_FLOAT_TYPE = MethodTypeDesc.ofDescriptor("(F)Ljava/lang/StringBuilder;");
+        static final MethodTypeDesc APPEND_INT_TYPE = MethodTypeDesc.ofDescriptor("(I)Ljava/lang/StringBuilder;");
+        static final MethodTypeDesc APPEND_LONG_TYPE = MethodTypeDesc.ofDescriptor("(J)Ljava/lang/StringBuilder;");
+        static final MethodTypeDesc APPEND_OBJECT_TYPE = MethodTypeDesc.ofDescriptor("(Ljava/lang/Object;)Ljava/lang/StringBuilder;");
+        static final MethodTypeDesc APPEND_STRING_TYPE = MethodTypeDesc.ofDescriptor("(Ljava/lang/String;)Ljava/lang/StringBuilder;");
+
+        static {
+            DUMPER = ClassFileDumper.getInstance("java.lang.invoke.StringConcatFactory.dump", "stringConcatClasses");
+        }
+
+        private SimpleStringBuilderStrategy() {
+            // no instantiation
+        }
+
+        private static MethodHandle generate(Lookup lookup, MethodType args, String[] constants) throws Exception {
+            String className = getClassName(lookup.lookupClass());
+
+            byte[] classBytes = ClassFile.of().build(ClassDesc.of(className),
+                    new Consumer<ClassBuilder>() {
+                        @Override
+                        public void accept(ClassBuilder clb) {
+                            clb.withFlags(AccessFlag.PUBLIC, AccessFlag.FINAL, AccessFlag.SUPER, AccessFlag.SYNTHETIC)
+                                .withMethodBody(METHOD_NAME,
+                                        MethodTypeDesc.ofDescriptor(args.toMethodDescriptorString()),
+                                        ClassFile.ACC_FINAL | ClassFile.ACC_PUBLIC | ClassFile.ACC_STATIC,
+                                        generateMethod(constants, args));
+                    }});
+
+            try {
+                Class<?> innerClass = lookup.defineHiddenClass(classBytes, true, STRONG).lookupClass();
+                DUMPER.dumpClass(className, innerClass, classBytes);
+                MethodHandle mh = lookup.findStatic(innerClass, METHOD_NAME, args);
+                return mh;
+            } catch (Exception e) {
+                DUMPER.dumpFailedClass(className, classBytes);
+                throw new StringConcatException("Exception while spinning the class", e);
+            }
+        }
+
+        private static Consumer<CodeBuilder> generateMethod(String[] constants, MethodType args) {
+            return new Consumer<CodeBuilder>() {
```

Comment:
Are you using inner classes for startup performance concerns? java.lang.invoke should be ready when you can reach here so using lambdas is safe.